src/Microsoft.AspNetCore.Mvc.Core/Internal/ControllerActionInvoker.cs

```diff
@@ -126,43 +81,10 @@ public class ControllerActionInvoker : IActionInvoker
 
                     var startTimestamp = _logger.IsEnabled(LogLevel.Information) ? Stopwatch.GetTimestamp() : 0;
 
-                    // The invoker is implemented using a 'Taskerator' or perhaps an 'Asyncerator' (both terms are correct
-                    // and in common usage). This method is the main 'driver' loop and will call into the `Next` method
-                    // (`await`ing the result) until a terminal state is reached.
-                    //
-                    // The `Next` method walks through the state transitions of the invoker and returns a `Task` when there's
-                    // actual async work that we need to await. As an optimization that Next method won't return a `Task`
-                    // that completes synchronously.
-                    //
-                    // Additionally the `Next` funtion will be called recursively when we're 'inside' a filter invocation.
-                    // Executing 'inside' a filter requires an async method call within a `try`/`catch` for error handling, so
-                    // we have to recurse. Each 'frame' calls into `Next` with a value of `Scope` that communicates what kind
-                    // of 'frame' is executing. This has an effect on the state machine transitions as well as what kinds of
-                    // contexts need to be constructed to communicate the result of execution of the 'frame'.
-
-                    // When returning, the `Next` method will set `next` to the state to goto on the subsequent invocation.
-                    // This is similar to `Task.ContinueWith`, but since we have a fixed number of states we can avoid
-                    // the overhead of actually using `Task.ContinueWith`.
-                    var next = State.InvokeBegin;
-
-                    // The `scope` tells the `Next` method who the caller is, and what kind of state to initialize to
-                    // communicate a result. The outermost scope is `Scope.Invoker` and doesn't require any type
-                    // of context or result other than throwing.
-                    var scope = Scope.Invoker;
-
-                    // The `state` is used for internal state handling during transitions between states. In practice this
-                    // means storing a filter instance in `state` and then retrieving it in the next state.
-                    var state = (object)null;
-
-                    // `isCompleted` will be set to true when we've reached a terminal state.
-                    var isCompleted = false;
-
                     try
                     {
-                        while (!isCompleted)
-                        {
-                            await Next(ref next, ref scope, ref state, ref isCompleted);
-                        }
+                        await InvokeFilterPipelineAsync();
```

Comment:
It is. We kick off the invocation from the subtype, and get called back at the end of resource filter invocation.